src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +36,208 @@ case class NoOpFingerprint() {
 }
 
 // IndexLogEntry-specific Content that uses IndexLogEntry-specific fingerprint.
-case class Content(root: String, directories: Seq[Content.Directory])
+case class Content(root: Directory, fingerprint: NoOpFingerprint = NoOpFingerprint()) {
+  // List of fully qualified paths of all files mentioned in this Content object.
+  @JsonIgnore
+  lazy val files: Seq[Path] = {
+    // Recursively find files from directory tree.
+    def rec(prefixPath: Path, directory: Directory): Seq[Path] = {
+      val files = directory.files.map(f => new Path(prefixPath, f.name))
+      files ++ directory.subDirs.flatMap { dir =>
+        rec(new Path(prefixPath, dir.name), dir)
+      }
+    }
+
+    rec(new Path(root.name), root)
+  }
+}
+
 object Content {
-  case class Directory(path: String, files: Seq[FileInfo], fingerprint: NoOpFingerprint)
-  object Directory {
-    // modifiedTime is an Epoch time in milliseconds. (ms since 1970-01-01T00:00:00.000 UTC).
-    case class FileInfo(name: String, size: Long, modifiedTime: Long)
-    object FileInfo {
-      def apply(s: FileStatus): FileInfo =
-        FileInfo(s.getPath.toString, s.getLen, s.getModificationTime)
+
+  /**
+   * Create a Content object from a directory path by recursively listing its leaf files. All
+   * files from the directory tree will be part of the Directory.
+   *
+   * @param path Starting directory path under which the files will be considered part of the
+   *             Directory object.
+   * @param pathFilter Filter for accepting paths. The default filter is picked from spark
+   *                   codebase, which filters out files like _SUCCESS.
+   * @param throwIfNotExists Throws FileNotFoundException if path is not found. Else creates a
+   *                         blank directory tree with no files.
+   * @return Content object with Directory tree starting at root, and containing all leaf files
+   *         from "path" argument.
+   */
+  def fromDirectory(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false): Content =
+    Content(Directory.fromDirectory(path, pathFilter, throwIfNotExists))
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Content = Content(Directory.fromLeafFiles(files))
+}
+
+case class Directory(
+    name: String,
+    var files: Seq[FileInfo] = Seq(),
```

Comment:
I think we could do without `var` by doing something like the following.
```
def dfs(files: Seq[String]) {
  val name = getTopDir(files.head)
  val nextFiles = files.drop(name.length + SEPARATOR.length)
  val (leafFiles, subDirFiles) = nextFiles.partition(_.contain(SEPARATOR)) // can be done in groupByTopDir
  val groupedPaths: Seq[Seq[String]] = groupByTopDir(subDirFiles)
  val subDirs = groupedPaths.tail.map(dfs(_))
  Directory( name, files=leafFiles, SubDirs=subDirs)
}
```
This is just a suggestion and I'd like to know your opinion on this.

Reply:
thanks @sezruby , I like the general idea of dfs but there are some string manipulations here which i am not sure about. Also I think the current existing implementation is better for pref for creating this structure which I have mentioned below. But I am ok if others also like this idea and prefer this Implementation.

Could you modify this to a `Path` based implementation?

Some doubts regarding perf:

1. `files.drop(name.length + SEPARATOR.length)` => Do you mean files.map(_.drop(name.length + SEPARATOR.length))`?
If that is the case, there will be a lot more string operations than we would like (lot of iterations over all file lists, removing the parent path one at a time). I guess this would hit performance very badly. (let me know if my understanding is wrong).
This would be (NM^2.log(N)) (M = file name length, N = No. of files)

2. `val (leafFiles, subDirFiles) = nextFiles.partition(_.contain(SEPARATOR))`: This is also bad for perf because at every level in the directory tree, we are iterating over full file list to partition it into leafFiles and subdirs.
This would be (NM)Log(NM) (N = no. of files, M = file name length)

3. String manipulation is something I don't personally like `(drop(n))` but I would be ok if the overall implementation is better for perf.

The current implementation is O (NM) (N = no. of files, M = file length) so i think it's better perf wise. If var vs val is a major concern, we can explore changing `Seq` to `ArrayBuffer` so that it's expandable and mutable.

Reply:
1. Yes it's `files.map(_.drop(name.length + SEPARATOR.length))`.
2. This can be done with groupByTopDir. 
3. I agree that string manipulation might be expensive and uncomfortable. But it seems Path.getParent also involves string operation. [github](https://github.com/apache/hadoop/blob/139a43e98e24320aeba23064fe53dc7055db9c4a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java#L426)

Regarding performance, I think this could be done in a more efficient way by using sorted file paths & `prefixLen` for common prefix. (though I'm not sure it's practical in Scala)

Reply:
if it's mainly for `val` vs `var`, I would suggest we can explore using either a `Builder()` pattern or use `ArrayBuffers()` instead of immutable `Seq()`. In both of those cases we can stick with `val` and still get the performance of the existing implementation. 

The one problematic issue with the suggestion is that it's really depth dependent. If all files are at depth 50, then we iterate over _all_ files 50 times. This makes me feel the benefit of `val` doesn't outweigh the cost of perf.

It's possible that I might have not fully understood the implementation. Could you please add an implementation/psuedocode of `groupByTopDir`

Reply:
Actually It's not all files. Each file was iterated by groupBY [its depth] times as they are partitioned during the traversal. In the edited version, there is no groupedTopDir, but just groupBy with a mapping function. 

'''
 val groupedPaths [string, seq[string]] = files.groupBy(p => p.slice(prefixLen, p.indexOf('/', prefixLen))
'''

We can use sorting of filePaths to reduce the cost of groupBy, but I think the edited ver. is not that bad.

Anyway, it's okay to keep cur ver. if it's preferable. 

cc. @imback82, @pirz could you give a comment for this?

Reply:
Sorry for the delay. I will get to this thread this weekend.

Reply:
I am +1 for using `val` for this case class.

Regarding the perf, we can address them in a separate PR with a proper benchmark if needed.

Reply:
Thanks @imback82 ,
Edited the code to use `ListBuffer` instead of `var`. This way we can still avoid unnecessary iterations over file list and keep the `val`. Please Let me know if this is ok or if dfs is preferred.

Reply:
The current approach looks good to me.